### PR TITLE
Add and use abstractions over labels.Labels

### PR DESCRIPTION
**Pull request by bboreham**
This is #10887, adding all the abstractions but not actually changing `labels.Labels`.
This should be possible to merge without breaking too much downstream.

There are one or two breaking changes, like `IndexReader.Series` gets a new `ScratchBuilder` parameter.

### CODE REVIEW

**Comment by roidelapluie** [comment]
https://github.com/prometheus/prometheus/blob/6a8702804e1116c8adf230b977aea3f55b522ce6/model/labels/labels.go#L544

This should become IsEmpty because we often call it with labels.EmptyLabels() which is not nil

File: scrape/scrape.go [hunk 1]
```diff
@@ -720,15 +721,24 @@ func resolveConflictingExposedLabels(lb *labels.Builder, exposedLabels, targetLa
 	}
 }
 
+func labelSliceHas(lbls []labels.Label, name string) bool {
+	for _, l := range lbls {
+		if l.Name == name {
+			return true
+		}
+	}
```
Comment (roidelapluie) [inline scrape/scrape.go:729]: does it make sense to assume sorted and do something like `if l.Name > name {return false}` ?

**Review by roidelapluie — APPROVED** [review]
LGTM

File: scrape/scrape.go (hunk 1 above)
Comment (bboreham) [reply scrape/scrape.go:729]: I think not: we know `conflictingExposedLabels` is sorted at the beginning, but after we have renamed one `foo` to `exported_foo` it is no longer certain.

**Comment by bboreham** [comment]
I changed `IndexReader.Series()` again - now it only takes a `ScratchBuilder`, and the caller can call `Labels()` on that if they want.

This is simpler, more foolproof, and faster in a couple of places where the labels value is not used.

**Comment by bboreham** [comment]
@roidelapluie, @bwplotka (since you reviewed related PRs) I am ready to merge this; please let me know if I should wait a little (e.g. until 2.41 is released).

**Comment by roidelapluie** [comment]
You can merge this, 2.41 is on its branch.

**Comment by bwplotka** [comment]
Feel free to go 👍🏽

**Pull request merged**